### DISCUSSION TITLE
Testing React components with WAGMI hooks

### DISCUSSION THREAD

**Discussion by calebtuttle** [wevm/wagmi#424] [Q&A]
Hi, 

I am using Jest to test some React components that make use of WAGMI hooks, specifically `useAccount()` and `useSigner()`. The tests always fail during rendering because the return value of these hooks is undefined. The hooks work as expected when I run my app regularly, but they don't seem to be pulling/returning data from the provider in the test environment. Is there a step I might be missing in setting up my tests? How can I test React components that include values from WAGMI hooks?

The following is my `createClient()` config and the `render()` call that fails.

    // copied from wagmi/packages/react/test/utils.ts
    const client = createClient({
      connectors: [
        new MockConnector({
          options: {
            signer: getSigners()[0],
          },
        }),
      ],
      provider: getProvider,
    });

    render(
      <Provider client={client}>
          <Element />
      </Provider>
    );

**Comment by pavel-lens** [comment]
Do you wait for account to connect?

```ts
// utils.ts
function useAccountWithConnectAndDisconnect(config: UseAccountConfig = {}) {
  return {
    account: useAccount(config),
    connect: useConnect(),
    disconnect: useDisconnect(),
  };
}

// component.test.ts
it('connects to WAGMI and renders correctly connected wallet address', async () => {
    const utils = renderHook(() => useAccountWithConnectAndDisconnect());
    const { result, waitFor } = utils;

    await actConnect({ utils });

    await waitFor(() => expect(result.current.account.isConnected).toBeTruthy());

    // console.log(result.current.account);
    const component = render(
      wrapper({
        children: <div>{result.current.account.address}</div>,
      })
    );

    expect(component).not.toBeUndefined();
    expect(component).toMatchSnapshot();
  });
```

Have a look in the following files for reference of `useAccount()`, `useConnect()` and `useDisconnect()` functions which are used in the first function.

https://github.com/wagmi-dev/wagmi/blob/main/packages/react/test/utils.ts
